File: 백준/Gold/14567. 선수과목 （Prerequisite）/선수과목 （Prerequisite）.cc
// Authored by : chjh2129
#include <bits/stdc++.h>

using namespace std;

int n, m;
vector<int> adj[1005]; // 인접 리스트 
int deg[1005]; // 정점의 indegree 값 
int semester[1005]; // 각 과목을 이수할 수 있는 학기 정보

void topological_sort(){
  queue<int> q;
  
  for(int i = 1; i <= n; i++){
    if(deg[i]) continue;
    semester[i] = 1;
    q.push(i);
  }

  // 위상 정렬을 진행하면서 이수할 수 있는 학기 정보를 갱신합니다.
  while(q.size()){
    auto cur = q.front(); q.pop();
    for(int nxt : adj[cur]){
      deg[nxt]--;
      if(deg[nxt] > 0) continue;
      semester[nxt] = semester[cur] + 1;
      q.push(nxt);
    }
  }
}

int main(){
  cin.tie(nullptr)->sync_with_stdio(false);

  // input
  cin >> n >> m;
  while(m--){
    int u, v;
    cin >> u >> v;
    adj[u].push_back(v); // 간선 삽입
    deg[v]++; // 정점 v의 indegree 값 증가 
  }

  // solve  
  topological_sort();

  // output
  for(int i = 1; i <= n; i++) cout << semester[i] << ' ';
}

File: 백준/Gold/14567. 선수과목 （Prerequisite）/선수과목 （Prerequisite）.java
import java.io.*;
import java.util.*;

public class Main {
  
  static int N, M;
  static List<List<Integer>> adj = new ArrayList<>();
  static int[] deg;
  static int[] semester;
  static Queue<Integer> Q = new LinkedList<>();

  public static void main(String[] args) throws IOException {
    BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    StringBuilder sb = new StringBuilder();
    StringTokenizer st = new StringTokenizer(br.readLine());

    // input & init
    N = Integer.parseInt(st.nextToken());
    M = Integer.parseInt(st.nextToken());

    for(int i = 0; i <= N; i++) adj.add(new ArrayList<>());
    deg = new int[N+1];
    semester = new int[N+1];

    while(M-- >0){
      st = new StringTokenizer(br.readLine());
      int u = Integer.parseInt(st.nextToken());
      int v = Integer.parseInt(st.nextToken());

      adj.get(u).add(v);
      deg[v]++;
    }

    // topological sort
    for(int i = 1; i <= N; i++){
      if(deg[i] > 0) continue;
      
      Q.add(i);
      semester[i] = 1;
    }

    while(!Q.isEmpty()){
      var cur = Q.poll();
      for(var nxt : adj.get(cur)){
        deg[nxt]--;
        if(deg[nxt] > 0) continue;
        semester[nxt] = semester[cur] + 1;
        Q.add(nxt);
      }
    }

    for(int i = 1; i <= N; i++){
      sb.append(semester[i]+" ");
    }

    // output
    System.out.println(sb.toString());
  }
}